src/main/kotlin/nonblocking/ReactorSelectorManager.kt

```diff
@@ -0,0 +1,89 @@
+package nonblocking
+
+import kotlinx.coroutines.*
+import java.io.Closeable
+import java.nio.channels.CancelledKeyException
+import java.nio.channels.SelectableChannel
+import java.nio.channels.SelectionKey
+import java.nio.channels.Selector
+import java.util.concurrent.ConcurrentHashMap
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+class ReactorSelectorManager : CoroutineScope, Closeable {
+    override val coroutineContext: CoroutineContext = Dispatchers.Default + CoroutineName("ReactorSelectorManager")
+
+    private val selector: Selector = Selector.open()
+
+    private val job: Job = launch {
+        while (true) {
+            selector.select()
+            val selectionKeys = selector.selectedKeys().iterator()
+            while (selectionKeys.hasNext()) {
+                val key = selectionKeys.next()
+                selectionKeys.remove()
+                try {
+                    if (!key.isValid) continue
+
+                    val attachment = key.attachment() as Attachment
+                    attachment.resumeContinuation(key)
+                } catch (e: CancelledKeyException) {
+                    key.channel().close()
+                }
+            }
+        }
+    }
+
+    suspend fun select(selectionKey: SelectionKey, interest: Int) {
+        suspendCoroutine { continuation ->
+            (selectionKey.attachment() as Attachment).apply {
+                addContinuation(interest, continuation)
+            }
+        }
+    }
+
+    fun addInterest(channel: SelectableChannel, interest: Int): SelectionKey {
+        val result = channel.keyFor(selector)?.apply {
+            interestOpsOr(interest)
+        } ?: let {
+            channel.register(selector, interest, Attachment())
+        }
+        selector.wakeup()
+        return result
+    }
+
+    fun deleteInterest(selectionKey: SelectionKey, interest: Int) {
+        selectionKey.interestOpsAnd(interest.inv())
+    }
+
+    override fun close() {
+        job.cancel()
+        job.invokeOnCompletion { selector.close() }
+        coroutineContext.cancel()
+    }
+}
+
+private class Attachment {
+    val interestToContinuationMap: ConcurrentHashMap<Int, Continuation<Unit>> = ConcurrentHashMap()
```

Comment:
We have 4 interests. Let's drop a map and make 4 fields instead. Let's also discuss if this case is concurrent

Reply:
Yep, I dropped a mad and did a 4 field of type `Continuation<Unit>?`, I also checked this case isn't concurrent